File: worlds-e2e/src/selenium/applicationPageModel.js
// @flow

import { WebDriver, By, until } from 'selenium-webdriver';
import HeaderBarPageModel from './headerBarPageModel';

/** Locator for the application header bar */
const HEADER_BAR_LOCATOR = By.css('[data-test="headerBar"]');

/**
 * Page Model representing the entire application
 */
export default class ApplicationPageModel {
    /** The web driver representing the application */
    _webDriver: WebDriver;

    /**
     * Construct the page model
     * @param webdriver The web driver that represents the page
     */
    constructor(webdriver: WebDriver) {
        this._webDriver = webdriver;
    }

    /**
     * Get the header bar page model
     * @return {HeaderBarPageModel} the header bar page model
     */
    async getHeaderBar(): Promise<HeaderBarPageModel> {
        await this._webDriver.wait(until.elementLocated(HEADER_BAR_LOCATOR));
        const headerBarElement = await this._webDriver.findElement(HEADER_BAR_LOCATOR);
        return new HeaderBarPageModel(headerBarElement);
    }
}

File: worlds-e2e/src/main/kotlin/uk/co/grahamcox/worlds/e2e/browser/ApplicationPageModel.kt
package uk.co.grahamcox.worlds.e2e.browser

import org.assertj.core.api.Assertions
import org.openqa.selenium.WebElement
import org.openqa.selenium.support.FindBy
import org.openqa.selenium.support.PageFactory
import org.openqa.selenium.support.pagefactory.DefaultElementLocatorFactory

/**
 * Page model representing the entire application
 */
class ApplicationPageModel(element: WebElement) {
    init {
        PageFactory.initElements(DefaultElementLocatorFactory(element), this)
    }

    /** The Web Element that represents the Header Bar */
    @FindBy(css = "[data-test=headerBar]")
    private lateinit var headerBarElement: WebElement

    /** Get the header bar of the page */
    val header: HeaderBarPageModel
        get() {
            Assertions.assertThat(headerBarElement.isDisplayed)
                    .`as`("The header bar should be visible")
                    .isTrue()
            return HeaderBarPageModel(headerBarElement)
        }

}
